### PR TITLE
CLN/TST: Remove nan issues with NumPy

### PR DESCRIPTION
**Pull request by bashtage**
Clean up nans that are being passsed to NumPy and SciPy
Remove calls to print in tests

### CODE REVIEW

**Comment by josef-pkt** [comment]
this contains several refactorings and nan enhancements that need review, it's not just TST

File: statsmodels/tools/tools.py [hunk 1]
```diff
@@ -354,24 +354,37 @@ def pinv_extended(X, rcond=1e-15):
...
     equal to 0 to 0. Therefore, it presumes that X should be positive in
     general.
     """
-    x = np.maximum(np.asarray(X).astype(np.float64), 0)
-    return np.greater(x, 0.) / (x + np.less_equal(x, 0.))
 
+    x = np.asarray(x)
+    out = np.zeros_like(x, dtype=np.float64)
+    nans = np.isnan(x.flat)
+    pos = np.logical_not(nans)
```
Comment (jbrockmendel) [inline statsmodels/tools/tools.py]: Is there an agreed-upon convention in favor of `logical_not(foo)` over `~foo`?

**Comment by bashtage** [comment]
I don't think so - there are 5 uses of logical_not .

File: statsmodels/nonparametric/tests/test_kernels.py [hunk 2]
```diff
@@ -19,10 +19,10 @@
 results = np.recfromcsv(os.path.join(curdir, fname))
 
 y = results['accident']
-x = np.log(results['service'])
-use_mask = ~np.isnan(x)
-x = x[use_mask]
-y = y[use_mask]
+x = results['service']
+positive = x >= 0
+x = np.log(x[positive].copy())
```
Comment (jbrockmendel) [inline statsmodels/nonparametric/tests/test_kernels.py]: why is the `copy()` needed?

File: statsmodels/stats/contrast.py [hunk 3]
```diff
@@ -55,7 +55,10 @@ def __init__(self, t=None, F=None, sd=None, effect=None, df_denom=None,
                 self.pvalue = self.dist.sf(self.statistic, df_denom)
             else:
                 "normal"
-                self.pvalue = self.dist.sf(np.abs(value)) * 2
+                pvalue = np.full_like(value, np.nan)
+                not_nan = ~np.isnan(value)
+                pvalue[not_nan] = self.dist.sf(np.abs(value[not_nan])) * 2
+                self.pvalue = pvalue
```
Comment (jbrockmendel) [inline statsmodels/stats/contrast.py]: No objection to be explicit and careful, but are there cases where the 1-line version messes up?  My n=1 test looked the same.

**Comment by jbrockmendel** [comment]
Looks like appveyor error is the old recarray problem.  I think that has been resolved (or silenced), so that should go away with a CI re-run.

File: statsmodels/nonparametric/tests/test_kernels.py (hunk 2 above)
Comment (bashtage) [reply statsmodels/nonparametric/tests/test_kernels.py]: It is not.

File: statsmodels/stats/contrast.py (hunk 3 above)
Comment (bashtage) [reply statsmodels/stats/contrast.py]: What is the 1 line version?

Comment (jbrockmendel) [reply statsmodels/stats/contrast.py]: GH can't find the diff anymore, but I think I was referring to the status-quo version.

**Comment by bashtage** [comment]
When a nan comes up it triggers a warning.  Can't run nans through sf

**Comment by jbrockmendel** [comment]
Makes sense, thanks for clearing that up.

+1

**Comment by jbrockmendel** [comment]
Travis error is unrelated test_arima111_predict_exog_2127.  This looks good to go.

**Comment by jbrockmendel** [comment]
@josef-pkt 5 dollar bill right here on the sidewalk, nearing its first birthday.

**Comment by bashtage** [comment]
@josef-pkt Still ready to be merged. Silences a lot of the noise in test runs

**Comment by josef-pkt** [comment]
merged
I'm not sure we are supposed to have nans at a few places like _calc_survfunc_right or whether it indicates a different problem.

I changed the merge message because it's not just CLN. I don't have a good overview where we want to allow for nans that are internally produced. I think we never really consistently reviewed this.
An alternative to avoiding the scipy distribution noise that has been showing up for some time now, would be to use catch warnings. I guess that in the distribution case the nan check is duplicate, once by statsmodels and once by scipy.stats. (But I don't remember the details or whether that has changed in the scipy.stats.distributions.)

@bashtage @jbrockmendel  thank you

**Pull request merged**